mod.d/speculativeexecution.yaml

```diff
@@ -0,0 +1,196 @@
+              spectrev2 = spectrev2_file.read()
+              print(spectrev2)
+              if re.search("retpoline", spectrev2):
+                  print("Retpoline is in use to mitigate Spectre V2\n")
+                  return True
+              elif re.search("Vulnerable: Retpoline", spectrev2):
+                      print("Retpoline is in use to mitigate Spectre V2, but IBPB is not showing as enabled.\n")
+                      return True
+              else:
+                  print("Retpoline is not in use to mitigate Spectre V2\n")
+                  return False
+      except Exception:
+          print("[WARN] Unable to determine Spectre V2 mitigation status\n")
+          raise
+
+  def get_pcid_status():
+      """
+      Determine if the system supports pcid or nvpcid. Pre-Haswell CPUs should only have PCID support.
+
+      Returns:
+          pcid (string): Highest PCID implementation supported
+      """
+      try:
+          with open("/proc/cpuinfo", "r") as cpuinfo_file:
+              if re.search("invpcid", cpuinfo_file.read()):
+                  print("invpcid supported. This is the best case for Kernel Page Table Isolation performance.\n")
+                  pcid = "invpcid"
+                  return pcid
+              elif re.search("pcid", cpuinfo_file.read()):
+                  print("pcid supported. This is a moderate case for Kernel Page Table Isolation performance.\n")
+                  pcid = "pcid"
+                  return pcid
+              else:
+                  print("No pcid support detected. This is the worst case for Kernel Page Table Isolation performance.\n")
+                  pcid = "none"
+      except Exception:
+          print("[WARN] Unable to determine pcid support\n")
+          raise
+
+  def run():
+      """
+      Determines mitigation status for the speculative execution exploits as well as expected performance
+      characteristics.
+
+      Returns:
+          True (bool): If collection is successfull.
+      """
+      try:
+          kpti = False
+          if is_intel_cpu():
+              print("Continuing on to determine Meltdown related status\n")
+              if get_kpti_status():
+                  kpti = True
+              get_pcid_status()
+          else:
+              print("Skipping Meltdown related collection.\n")
+
+          if get_retpoline_status():
+              print("Retpoline is a high performance mitigation for Spectre v2 and should not result in negative"
+              "performance impact.\n")
+
+          if kpti:
+              print("Kernel Page Table Isolation can result in negative performance impact to certain workloads.\n")
+              print("We recommend "
+              "visiting https://aws.amazon.com/premiumsupport/knowledge-center/ec2-linux-performance/ for ways you can "
+              "increase EC2 instance performance.")
+
+          return True
+      except Exception as ex:
```

Comment:
There should probably be exception handling for a case where a file does not exist and the check function allows the exception to be bubbled up. For example, if run on a system with an old kernel, do we want to module to generate a warning or just state "hey we can't check this"? Thankfully, it [looks like](https://lore.kernel.org/patchwork/cover/870579/) there is no scenario where only one of the two checked files exists.

Reply:
Since it's not a diagnostic module I don't know that there are any concerns in it generating warnings.